plugin/plugin.go

```diff
@@ -8,68 +8,65 @@ import (
 	"context"
 	"errors"
 	"fmt"
+	"io"
+	"strings"
 
 	"github.com/drone/drone-go/plugin/validator"
 	"github.com/open-policy-agent/opa/rego"
-	"gopkg.in/yaml.v2"
+	yaml "gopkg.in/yaml.v2"
 )
 
-// New returns a new validator plugin.
 func New(policy string) validator.Plugin {
 	return &plugin{
 		policyPath: policy,
 	}
 }
 
-type plugin struct {
-	policyPath string
-}
-
 func (p *plugin) Validate(ctx context.Context, req *validator.Request) error {
-	// Drone sends its raw config in the request body
-	type DroneConfig struct {
-		Kind  string `yaml:"kind" json:"kind"`
-		Type  string `yaml:"type" json:"type"`
-		Name  string `yaml:"name" json:"name"`
-		Steps []struct {
-			Name      string   `yaml:"name" json:"name"`
-			Image     string   `yaml:"image" json:"image"`
-			Commands  []string `yaml:"commands" json:"commands"`
-			DependsOn []string `yaml:"depends_on,omitempty" json:"depends_on,omitempty"`
-		} `yaml:"steps" json:"steps"`
-		Trigger struct {
-			Branch []string `yaml:"branch" json:"branch"`
-		} `yaml:"trigger" json:"trigger"`
-	}
 
-	var droneConfig DroneConfig
+	var documents []DroneConfig
 
-	err := yaml.Unmarshal([]byte(req.Config.Data), &droneConfig)
-	if err != nil {
-		return err
+	dec := yaml.NewDecoder(strings.NewReader(req.Config.Data))
+	for {
+		var document DroneConfig
+
+		err := dec.Decode(&document)
+		if err == io.EOF {
+			break
+		} else if err != nil {
+			return err
+		}
+		documents = append(documents, document)
 	}
 
-	r := rego.New(
+	r, err := rego.New(
 		rego.Query("deny = data.drone.validation.deny; msg = data.drone.validation.out"),
-		rego.Load([]string{p.policyPath}, nil))
-
-	query, err := r.PrepareForEval(ctx)
+		rego.Load([]string{p.policyPath}, nil)).PrepareForEval(ctx)
 	if err != nil {
 		return err
 	}
 
-	rs, err := query.Eval(ctx, rego.EvalInput(droneConfig))
-	if err != nil {
-		return err
-	}
+	var message string
+	var values []string
 
-	if rs[0].Bindings["deny"] == true {
-		message := fmt.Sprintf("[validator] %v", rs[0].Bindings["msg"])
-	 	return errors.New(message)
+	for _, resource := range documents {
+		rs, err := r.Eval(ctx, rego.EvalInput(resource))
+		if err != nil {
+			return err
+		}
+		if rs[0].Bindings["deny"] == true {
+			s := strings.Split(fmt.Sprintf("%v", rs[0].Bindings["msg"]), ":")
```

Comment:
This relies heavily on the formatting of the default policy output string. I don't think the code should make these kinds of assumptions on what the policy look slike. Further, if a user were to remove the `:` from the output, an index out of bounds panic is triggered by the `s[1]` call.

I think you can probably just collect the output messages as a newline delimited string. However, you'll want to verify how this output looks in the drone UI.

Reply:
Indeed, that's one of the main reasons why I'm not sure about grouping messages from multiple pipeline documents.

Drone UI doesn't look good if we return multiple lines with different error messages due to how the error is presented.

Another point that I'd like to point is in terms of flexibility, if we decide that we want to start validating other components of the pipeline with an OPA policy that can return different messages according to the validation error, returning all messages at once would be pretty confuse to the end user (something like a box with multiple error sentences, one after the other). If we return the failure immediately, the user will be able to correlate the error message with the failed step, IMO simplifying and clarifying the troubleshooting process at the cost of fixing different errors individually after each build execution.

I'm keen to remove the latest changes and return errors as soon as they are parsed due to these reasons, what do you and @dskatz think?

Reply:
I think for the first iteration, returning each error as-is is OK.  Then we can watch and improve from there.  

If there is any parsing of messages returned from OPA, I think we should update OPA to return a serialized JSON string. This way the golang is much clearer in how that message is marshaled into an object.

Reply:
I'm fine with reverting for now and adding a TODO to improve later.

Reply:
Cool, commit reverted and TODO reference added.